solr/core/src/java/org/apache/solr/schema/ScalarQuantizedDenseVectorField.java

```diff
@@ -0,0 +1,154 @@
+   * spec
+   */
+  private int bits;
+
+  /**
+   * Confidence interval to use for scalar quantization Default is calculated as
+   * `1-1/(vector_dimensions + 1)`
+   */
+  private Float confidenceInterval;
+
+  /**
+   * When enabled, in conjunction with 4 bit size, will pair values into single bytes for 50%
+   * reduction in memory usage (comes at the cost of some decode speed penalty)
+   */
+  private boolean compress;
+
+  public ScalarQuantizedDenseVectorField() {
+    super();
+  }
+
+  public ScalarQuantizedDenseVectorField(
+      int dimension,
+      VectorSimilarityFunction similarityFunction,
+      VectorEncoding vectorEncoding,
+      int bits,
+      Float confidenceInterval,
+      boolean compress) {
+    super(dimension, similarityFunction, vectorEncoding);
+    this.bits = bits;
+    this.confidenceInterval = confidenceInterval;
+    this.compress = compress;
+  }
+
+  @Override
+  public void init(IndexSchema schema, Map<String, String> args) {
+    this.bits = ofNullable(args.remove(BITS_PARAM)).map(Integer::parseInt).orElse(DEFAULT_BITS);
+
+    this.compress =
+        ofNullable(args.remove(COMPRESS_PARAM)).map(Boolean::parseBoolean).orElse(false);
+
+    this.confidenceInterval =
+        ofNullable(args.remove(CONFIDENCE_INTERVAL_PARAM))
+            .map(Float::parseFloat)
+            .orElse(DEFAULT_CONFIDENCE_INTERVAL);
+
+    if (ofNullable(args.remove(DYNAMIC_CONFIDENCE_INTERVAL_PARAM))
+        .map(Boolean::parseBoolean)
+        .orElse(false)) {
+      this.confidenceInterval = Lucene99ScalarQuantizedVectorsFormat.DYNAMIC_CONFIDENCE_INTERVAL;
+    }
+
+    super.init(schema, args);
+  }
+
+  @Override
+  public KnnVectorsFormat buildKnnVectorsFormat() {
+    return new Lucene99HnswScalarQuantizedVectorsFormat(
+        getHnswMaxConn(),
+        getHnswBeamWidth(),
+        DEFAULT_NUM_MERGE_WORKER,
+        getBits(),
+        useCompression(),
+        getConfidenceInterval(),
+        null);
+  }
+
+  @Override
+  public void checkSchemaField(final SchemaField field) throws SolrException {
+    super.checkSchemaField(field);
+
+    if (confidenceInterval != null
+        && confidenceInterval != Lucene99ScalarQuantizedVectorsFormat.DYNAMIC_CONFIDENCE_INTERVAL
+        && (confidenceInterval < MINIMUM_CONFIDENCE_INTERVAL
+            || confidenceInterval > MAXIMUM_CONFIDENCE_INTERVAL)) {
+      throw new SolrException(
+          SolrException.ErrorCode.SERVER_ERROR,
+          getClass().getSimpleName()
+              + " fields must have non-dynamic confidence interval between 0.9 and 1.0 "
+              + field.getName());
+    }
+    if (getBits() != 4 && getBits() != 7) {
+      throw new SolrException(
+          SolrException.ErrorCode.SERVER_ERROR,
+          getClass().getSimpleName()
+              + " fields must have bit size of 4 (half-byte) or 7 (signed-byte) "
+              + field.getName());
+    }
+
+    if (useCompression() && getBits() != 4) {
```

Comment:
Can't we rely on "if (bits > 4 && compress) {
      // compress=true otherwise silently does nothing when bits=7?
      throw new IllegalArgumentException("compress=true only applies when bits=4");
    }" Lucene side?